providers/implementations/kem/mlx_kem.c

```diff
@@ -0,0 +1,338 @@
+     * same (TLS client) ECDHE keypair for both the classical EC keyshare and a
+     * corresponding ECDHE + ML-KEM keypair.  But the TLS layer would then need
+     * know that this is a hybrid, and that it can partly reuse the same keys
+     * as another group for which a keyshare will be sent.  Deferred until we
+     * support generating multiple keyshares, there's a workable keyshare
+     * prediction specification, and the optimisation is justified.
+     */
+    cbuf = ctext + (1 - ml_kem_slot) * key->minfo->ctext_bytes;
+    encap_clen = key->xinfo->pubkey_bytes;
+    ctx = EVP_PKEY_CTX_new_from_pkey(key->libctx, key->xkey, key->propq);
+    if (ctx == NULL
+        || EVP_PKEY_keygen_init(ctx) <= 0
+        || EVP_PKEY_keygen(ctx, &xkey) <= 0
+        || EVP_PKEY_get_octet_string_param(xkey, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY,
+                                           cbuf, encap_clen, &encap_clen) <= 0)
+        goto end;
+    if (encap_clen != key->xinfo->pubkey_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR,
+                       "unexpected %s public key output size: %lu",
+                       key->xinfo->algorithm_name, (unsigned long) encap_clen);
+        goto end;
+    }
+    EVP_PKEY_CTX_free(ctx);
+
+    /* Derive the ECDH shared secret */
+    encap_slen = key->xinfo->shsec_bytes;
+    sbuf = shsec + (1 - ml_kem_slot) * ML_KEM_SHARED_SECRET_BYTES;
+    ctx = EVP_PKEY_CTX_new_from_pkey(key->libctx, xkey, key->propq);
+    if (ctx == NULL
+        || EVP_PKEY_derive_init(ctx) <= 0
+        || EVP_PKEY_derive_set_peer(ctx, key->xkey) <= 0
+        || EVP_PKEY_derive(ctx, sbuf, &encap_slen) <= 0)
+        goto end;
+    if (encap_slen != key->xinfo->shsec_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR,
+                       "unexpected %s shared secret output size: %lu",
+                       key->xinfo->algorithm_name, (unsigned long) encap_slen);
+        goto end;
+    }
+
+    ret = 1;
+  end:
+    EVP_PKEY_free(xkey);
+    EVP_PKEY_CTX_free(ctx);
+    return ret;
+}
+
+static int mlx_kem_decapsulate(void *vctx, uint8_t *shsec, size_t *slen,
+                               const uint8_t *ctext, size_t clen)
+{
+    MLX_KEY *key = ((PROV_MLX_KEM_CTX *) vctx)->key;
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *xkey = NULL;
+    const uint8_t *cbuf;
+    uint8_t *sbuf;
+    size_t decap_slen = ML_KEM_SHARED_SECRET_BYTES + key->xinfo->shsec_bytes;
+    size_t decap_clen = key->minfo->ctext_bytes + key->xinfo->pubkey_bytes;
+    int ml_kem_slot = key->xinfo->ml_kem_slot;
+    int ret = 0;
+
+    if (!mlx_kem_have_prvkey(key)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
+    }
+
+    if (shsec == NULL) {
+        if (slen == NULL)
+            return 0;
+        *slen = decap_slen;
+        return 1;
+    }
+
+    /* For now tolerate newly-deprecated NULL length pointers. */
+    if (slen == NULL) {
+        slen = &decap_slen;
```

Comment:
Please do not tolerate it.

Reply:
Done, be it with some reservation about violating documented contract, even if already violated by other providers...